### PR TITLE
[CARBONDATA-1609] Thrift Changes to support Pre-aggregate

### PR DESCRIPTION
**Pull request by kumarvishal09**
Thrift changes done to support Pre-Aggregate

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1164/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/534/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1165/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/535/

File: format/src/main/thrift/schema.thrift [hunk 1]
```diff
@@ -165,9 +165,46 @@ struct TableSchema{
...
+enum RelationType {
+  AGGREGATION = 0;
+}
+struct RelationProperties {
+  1: optional RelationType relationTypes;
+  // to maintain properties like select query, query type like groupby, join
+  2: optional map<string, string> properties;
+}
+
+struct ChildSchema {
```
Comment (ravipesala) [inline format/src/main/thrift/schema.thrift]: Change as Datamap schema

File: format/src/main/thrift/schema.thrift [hunk 2]
```diff
@@ -165,9 +165,46 @@ struct TableSchema{
...
+  1: required RelationIdentifier relationInfo;
+  2: required TableSchema childTableSchema;
```
Comment (ravipesala) [inline format/src/main/thrift/schema.thrift]: Make it optional

File: format/src/main/thrift/schema.thrift [hunk 3]
```diff
@@ -165,9 +165,46 @@ struct TableSchema{
...
+  3: optional RelationProperties relationProperties;
+}
 struct TableInfo{
 	1: required TableSchema fact_table;
 	2: required list<TableSchema> aggregate_table_list;
+	3: optional list<ChildSchema> childSchemas; // childSchema information
+ // only in case of child table it will be filled to maintain the relation with parent tables
+   // to hanle operation like drop in that case parent table will be also updated
+  4: optional list<RelationIdentifier> parentRelationInfo;
```
Comment (ravipesala) [inline format/src/main/thrift/schema.thrift]: not required again

File: format/src/main/thrift/schema.thrift [hunk 4]
```diff
@@ -165,9 +165,46 @@ struct TableSchema{
...
+   1: required RelationIdentifier relationInfo;
+   2: required string columnId;
+   3: required string columnName
+}
+
+enum RelationType {
+  AGGREGATION = 0;
+}
+struct RelationProperties {
+  1: optional RelationType relationTypes;
```
Comment (ravipesala) [inline format/src/main/thrift/schema.thrift]: it is not required , but add the optional class name

File: format/src/main/thrift/schema.thrift [hunk 5]
```diff
@@ -165,9 +165,46 @@ struct TableSchema{
   4: optional map<string,string> tableProperties; // Table properties configured by the user
   5: optional BucketingInfo bucketingInfo; // Bucketing information
   6: optional PartitionInfo partitionInfo; // Partition information
+  7: optional list<ColumnRelation> columnRelationList;
```
Comment (ravipesala) [inline format/src/main/thrift/schema.thrift]: move this to datamap schema

**Comment by ravipesala** [comment]
just try this
```

struct RelationIdentifier {
   1: optional string databaseName;
   2: required string tableName;
   3: required string tableId;
}

struct ColumnRelation{
   // child column
   1: required ColumnSchema table_column;
   2: optional list<ParentColumnTableRelation> columnTableRelation;
}

struct ParentColumnTableRelation {
   1: required RelationIdentifier relationInfo;
   2: required string columnId;
   3: required string columnName
}


struct DataMapSchema {
  1: required string className;
  2: optional list<RelationIdentifier> relationInfos;
  3: optional list<ColumnRelation> columnRelationList;
  // to maintain properties like select query, query type like groupby, join
  4: optional map<string, string> properties;
}

struct TableInfo{
  1: required TableSchema fact_table;
  2: required list<TableSchema> aggregate_table_list;
  3: optional list<DataMapSchema> dataMapSchemas; // dataMapSchema information
}
```

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1170/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/539/

**Comment by kumarvishal09** [comment]
@ravipesala Updated thrift.
struct RelationIdentifier {
   1: optional string databaseName;
   2: required string tableName;
   3: required string tableId;
}

struct ParentColumnTableRelation {
   1: required RelationIdentifier relationIdentifier;
   2: required string columnId;
   3: required string columnName
}

struct DataMapSchema  {
    // class name
    1: required string className;
    // relation indentifier
    2: optional RelationIdentifier relationIdentifire;
    // in case of preaggregate it will be used to maintain the child schema
    // which will be usefull in case of query and data load
    3: optional TableSchema childTableSchema;
    // to maintain properties like select query, query type like groupby, join
    4: optional map<string, string> properties;
}

struct TableInfo{
	1: required TableSchema fact_table;
	2: required list<TableSchema> aggregate_table_list;
	3: optional list<DataMapSchema> dataMapSchemas; // childSchema information
}

Inside column schema added ParentColumnTableRelation

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/647/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1291/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1292/

File: core/src/main/java/org/apache/carbondata/core/memory/HeapMemoryAllocator.java [hunk 6]
```diff
@@ -17,12 +17,13 @@
 
 package org.apache.carbondata.core.memory;
 
-import javax.annotation.concurrent.GuardedBy;
 import java.lang.ref.WeakReference;
 import java.util.HashMap;
 import java.util.LinkedList;
 import java.util.Map;
 
+import javax.annotation.concurrent.GuardedBy;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/memory/HeapMemoryAllocator.java]: Not required to change

File: core/src/main/java/org/apache/carbondata/core/metadata/converter/SchemaConverter.java [hunk 7]
```diff
@@ -100,4 +101,13 @@ TableSchema fromExternalToWrapperTableSchema(
...
   TableInfo fromExternalToWrapperTableInfo(org.apache.carbondata.format.TableInfo externalTableInfo,
       String dbName, String tableName, String storePath);
+
+  /**
+   * method to convert thrift datamap schema object to wrapper
+   * data map object
+   * @param thriftchildSchema
+   * @return DataMapSchema
+   */
+  DataMapSchema fromExternalToWrapperChildSchema(
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/converter/SchemaConverter.java]: change method name to reflect datamap in it

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableInfo.java [hunk 8]
```diff
@@ -24,13 +24,19 @@
 import java.io.DataOutputStream;
 import java.io.IOException;
 import java.io.Serializable;
+import java.util.ArrayList;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableInfo.java]: use import java.util.*

**Comment by ravipesala** [comment]
Please change the branch to pre-aggregate

**Comment by ravipesala** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1298/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/659/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/863/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/951/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1566/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1006/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1623/

**Pull request closed without merging**